From b329d60f03bf5c882df9fa5f4e89f0fa5386d321 Mon Sep 17 00:00:00 2001
From: Ye Li <ye.li@nxp.com>
Date: Thu, 27 Jul 2023 09:52:33 +0800
Subject: [PATCH] LFU-573-2 imx8m: Reserve new IVT+CSF for FIT FDT signature

commit 5a0faefc223e51e088433663b6e7d6fbce89bf59 from
https://github.com/nxp-imx/imx-mkimage.git

Without using FIT FDT hash, we also allow user to sign FIT FDT structure,
so that FIT image can upgrade individually. The option needs
CONFIG_IMX_SPL_FIT_FDT_SIGNATURE enabled in SPL.

imx-mkimage will insert the new IVT for FIT FDT signature by default
and reserve the CSF (0x2000) for the FIT FDT signature.

Upstream-Status: Backport

Signed-off-by: Ye Li <ye.li@nxp.com>
---
 iMX8M/mkimage_imx8.c | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/iMX8M/mkimage_imx8.c b/iMX8M/mkimage_imx8.c
index 5d5343b..a829883 100644
--- a/iMX8M/mkimage_imx8.c
+++ b/iMX8M/mkimage_imx8.c
@@ -1031,6 +1031,24 @@ int generate_ivt_for_fit(int fd, int fit_offset, uint32_t ep, uint32_t *fit_load
 		exit(EXIT_FAILURE);
 	}
 
+	ret = lseek(fd, fit_offset + fit_size + CSF_SIZE, SEEK_SET);
+	if (ret < 0) {
+		fprintf(stderr, "%s: lseek error %s\n",
+				__func__, strerror(errno));
+		exit(EXIT_FAILURE);
+	}
+
+	flash_header_v2_t fdt_ivt_header = { { 0xd1, 0x2000, 0x40 },
+		load_addr, 0, 0, 0,
+		(load_addr + fit_size + CSF_SIZE ),
+		(load_addr + fit_size + CSF_SIZE + 0x20),
+		0 };
+
+	if (write(fd, &fdt_ivt_header, sizeof(flash_header_v2_t)) != sizeof(flash_header_v2_t)) {
+		fprintf(stderr, "FIT FDT IVT writing error on fit image\n");
+		exit(EXIT_FAILURE);
+	}
+
 	*fit_load_addr = load_addr;
 
 	return fit_offset + fit_size;
-- 
2.43.0

